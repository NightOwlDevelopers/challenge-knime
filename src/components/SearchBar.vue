<script setup lang="ts">
import type { PropType } from "vue";
import { Ref, computed, ref } from "vue";
import { WorkflowBlock } from "../types/dataTypes";

const props = defineProps({
  flowBlockList: {
    type: Object as PropType<WorkflowBlock[]>,
    default: [],
  },
});

const emit = defineEmits<{
  (e: "add-block", block: WorkflowBlock): void;
}>();

const searchQuery: Ref<string> = ref("");

const addBlock = (block: WorkflowBlock) => {
  emit("add-block", block);
};

const searchNodes = computed(() => {
  if (searchQuery.value) {
    const regExpQuery = new RegExp(searchQuery.value, "i");
    console.log(props.flowBlockList);
    return props.flowBlockList.filter(
      (node) => regExpQuery.test(node.name) || regExpQuery.test(node.name)
    );
  }
  return props.flowBlockList;
});
</script>

<template>
  <div class="search-bar">
    <div class="search-bar__wrapper">
      <input
        class="search-bar__input"
        v-model="searchQuery"
        type="text"
        placeholder="Add new node..."
      />
      <ul class="search-bar__items-list">
        <li
          class="search-bar__list_item"
          v-for="node in searchNodes"
          :key="node.name"
          :value="node.name"
          @mousedown="addBlock(node)"
        >
          <div>{{ node.name }}</div>
          <img :src="node.image" alt="" />
        </li>
      </ul>
    </div>
  </div>
</template>

<style scoped>
.search-bar {
  width: 100%;
  background-color: var(--vt-c-white);
  overflow: hidden;
  display: flex;
  justify-content: center;
  overflow: visible;
  padding: 2rem;
}

.search-bar__wrapper {
  width: 100%;
  max-width: 300px;
  display: flex;
  flex-direction: column;
  position: relative;
}

.search-bar__input {
  font-size: 1rem;
  height: 2rem;
  width: 100%;
  border: 1px solid;
  padding: 0.25rem;
  border-radius: 5px;
}
.search-bar__wrapper:focus-within .search-bar__items-list {
  visibility: visible;
}

.search-bar__items-list {
  list-style: none;
  margin: 0;
  width: 100%;
  background-color: var(--vt-c-white);
  display: block;
  position: absolute;
  top: 100%;
  visibility: hidden;
  z-index: 100;
  max-height: 200px;
}

.search-bar__items-list:hover {
  cursor: pointer;
}

.search-bar__list_item {
  font-size: 1rem;
  border: 1px solid var(--vt-c-white-mute);
  width: 100%;
  display: flex;
  justify-content: space-between;
  padding: 1px;
  z-index: 100;
}

img {
  height: 40px;
  width: 40px;
}
</style>
