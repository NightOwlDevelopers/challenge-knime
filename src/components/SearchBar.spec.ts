import { shallowMount, mount } from "@vue/test-utils";
import Vuex from "vuex";
import SearchBar from "../components/SearchBar.vue";

import { describe, it, expect } from "vitest";
import { workflowBlocksMock } from "../dataMock";

describe("SearchBar.vue", () => {
  it("renders correctly", async () => {
    const wrapper = mount(SearchBar, {
      propsData: {
        flowBlockList: workflowBlocksMock,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it("by default displays all available blocks are displays in the options list", () => {
    const wrapper = shallowMount(SearchBar, {
      propsData: {
        flowBlockList: workflowBlocksMock,
      },
    });

    wrapper.find("[test-id='search-bar--wrapper']").trigger("click");
    const options = wrapper.findAll("[test-id='search-bar--list-item']");

    expect(options.length).toEqual(workflowBlocksMock.length);
  });

  it("only block matching search query is displayed when search query entered", async () => {
    const wrapper = shallowMount(SearchBar, {
      propsData: {
        flowBlockList: workflowBlocksMock,
      },
    });
    await wrapper
      .find("[test-id='search-bar--input']")
      .setValue(workflowBlocksMock[0].name);

    const listItems = wrapper.findAll("[test-id='search-bar--list-item']");

    expect(listItems.length).toEqual(1);
  });

  it("emits add-block event with correct payload on selecting block", async () => {
    const wrapper = shallowMount(SearchBar, {
      propsData: {
        flowBlockList: workflowBlocksMock,
      },
    });
    await wrapper
      .find("[test-id='search-bar--input']")
      .setValue(workflowBlocksMock[0].name);

    wrapper.find("[test-id='search-bar--list-item']").trigger("mousedown");
    expect(wrapper.emitted()["add-block"]).toEqual([[workflowBlocksMock[0]]]);
  });
});
