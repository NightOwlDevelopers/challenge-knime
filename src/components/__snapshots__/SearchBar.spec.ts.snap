// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar.vue > renders correctly 1`] = `
<div
  class="search-bar"
  data-v-6ab8b910=""
>
  <div
    class="search-bar__wrapper"
    data-v-6ab8b910=""
    test-id="search-bar--wrapper"
  >
    <input
      class="search-bar__input"
      data-v-6ab8b910=""
      placeholder="Add new node..."
      test-id="search-bar--input"
      type="text"
    />
    <ul
      class="search-bar__items-list"
      data-v-6ab8b910=""
      test-id="search-bar--items-list"
    >
      
      <li
        class="search-bar__list_item"
        data-v-6ab8b910=""
        test-id="search-bar--list-item"
        value="0"
      >
        <div
          data-v-6ab8b910=""
        >
          Data generator
        </div>
        <img
          alt=""
          data-v-6ab8b910=""
          src="assets/data_generator.jpg"
        />
      </li>
      <li
        class="search-bar__list_item"
        data-v-6ab8b910=""
        test-id="search-bar--list-item"
        value="0"
      >
        <div
          data-v-6ab8b910=""
        >
          CASE Switch Data (Start)
        </div>
        <img
          alt=""
          data-v-6ab8b910=""
          src="assets/case_switch.jpg"
        />
      </li>
      <li
        class="search-bar__list_item"
        data-v-6ab8b910=""
        test-id="search-bar--list-item"
        value="0"
      >
        <div
          data-v-6ab8b910=""
        >
          Cluster Assigner
        </div>
        <img
          alt=""
          data-v-6ab8b910=""
          src="assets/cluster_assigner.jpg"
        />
      </li>
      
    </ul>
  </div>
</div>
`;
